# Cluster Analysis in Toronto

Webscraping a Wikipedia page and creating a datafarme using pandas. Adding Foursquare location data using latitude and longitude coordinates of each neighborhood. Cluster the neighbourhoods via k-means.

## Code

In [98]:
import pandas as pd
import pgeocode
import numpy as np
import folium
import requests
import json
from sklearn.cluster import KMeans


In [4]:
df = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')[0]

#### 1 Scrape Wikipedia
The first table on the Wikipedia page about Toronto Postal Codes is imported.

In [7]:
df_filtered = df[df['Borough'] != 'Not assigned'] 
df_filtered.reset_index(inplace = True)
df_filtered.drop(['index'], axis = 1, inplace = True)
df_filtered.rename(columns = {"Postal Code": "PostalCode"}, inplace = True)
print(df_filtered.head())

  PostalCode           Borough                                Neighbourhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government


C:\Users\maiks\anaconda3\lib\site-packages\pandas\core\frame.py:3997: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,
C:\Users\maiks\anaconda3\lib\site-packages\pandas\core\frame.py:4133: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Not assigned postal codes are dropped from the table. The index of the dataframe has been reset; the old index was dropped.

In [6]:
df_filtered.shape

(103, 3)

The shape of the table is 103 rows and 3 columns.

#### 2 Coordinates
Add geographical coordinates to each postal code.

In [9]:
ca = pgeocode.Nominatim('ca')
postal_codes = df_filtered['PostalCode'].tolist()
df_pgeocode = ca.query_postal_code(postal_codes)
#print(df_pgeocode)

Now clean the table and merge with 'df_filtered'.

In [10]:
df_coord = df_pgeocode[['postal_code', 'latitude', 'longitude']]
df_coord.rename(columns = {"postal_code": "PostalCode", "latitude": "Latitude", "longitude": "Longitude"}, inplace = True)
#print(df_coord)

C:\Users\maiks\anaconda3\lib\site-packages\pandas\core\frame.py:4133: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [19]:
df_merged = pd.merge(df_filtered, df_coord, on = 'PostalCode')
pd.set_option('display.max_rows', None)
print(df_merged.head())
print(df_merged[75:78])

  PostalCode           Borough                                Neighbourhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

   Latitude  Longitude  
0   43.7545   -79.3300  
1   43.7276   -79.3148  
2   43.6555   -79.3626  
3   43.7223   -79.4504  
4   43.6641   -79.3889  
   PostalCode       Borough  \
75        M6R  West Toronto   
76        M7R   Mississauga   
77        M9R     Etobicoke   

                                        Neighbourhood  Latitude  Longitude  
75                             Parkdale, Roncesvalles   43.6469   -79.4521  
76              Canada Post Gateway Processing Centre       NaN        NaN  
77  Kingsview Villa

There is a problem with the row with index 76. No coordinates are extracted. It turns out that this postal code is exclusively assigned to a branch of Canadian Post. The missing coordinates were therefore entered manually.   

In [20]:
df_merged["Latitude"].fillna(43.635060, inplace = True)
df_merged["Longitude"].fillna(-79.618030, inplace = True) 
print(df_merged)

    PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarbor

#### 3 Map Neighbourhoods
Create a map of Tornoto Cluster neighbourhoods of Toronto.

In [27]:
map_toronto = folium.Map(location=[43.653963, -79.387207], zoom_start=11)

for lat, long, post, borough, neighbourhood in zip(df_merged['Latitude'], df_merged['Longitude'], df_merged['PostalCode'], df_merged['Borough'], df_merged['Neighbourhood']):
    label = "{}, {}".format(neighbourhood, borough)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

In [29]:
unique_boroughs = df_merged['Borough'].unique()
print(unique_boroughs)

['North York' 'Downtown Toronto' 'Etobicoke' 'Scarborough' 'East York'
 'York' 'East Toronto' 'West Toronto' 'Central Toronto' 'Mississauga']


#### 4 Narrow down Neighbourhoods
There are 10 different boroughs in Toronto. I will focus on those which have 'Toronto' in their name.

In [33]:
borough_names = list(df_merged.Borough.unique())

borough_central_df = []

for x in borough_names:
    if "toronto" in x.lower():
        borough_central_df.append(x)
        
print(borough_central_df)

toronto_central_df = df_merged[df_merged['Borough'].isin(borough_central_df)].reset_index(drop=True)
print(toronto_central_df.shape)
toronto_central_df.head()
toronto_central_df.shape

['Downtown Toronto', 'East Toronto', 'West Toronto', 'Central Toronto']
(39, 5)


(39, 5)

In [34]:
map_toronto = folium.Map(location=[43.653963, -79.387207], zoom_start=11)

for lat, long, post, borough, neighbourhood in zip(toronto_central_df['Latitude'], toronto_central_df['Longitude'], toronto_central_df['PostalCode'], toronto_central_df['Borough'], toronto_central_df['Neighbourhood']):
    label = "{}, {}".format(neighbourhood, borough)
    popup = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, long],
        radius=5,
        popup=popup,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)
    
map_toronto

Use Foursquare API to explore the boroughs and segment them.

In [35]:
CLIENT_ID = 'BD2UK4A1FFBLLHLCWHTBQKJLCBVGWN5DCCKGW0X15EQM4ILT'
CLIENT_SECRET = 'GEZSONRDNNHU3QSLGFDOLYSV3YSIB5KKM1C2IDNUKIHEYJIL'
VERSION = '20200731'

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: BD2UK4A1FFBLLHLCWHTBQKJLCBVGWN5DCCKGW0X15EQM4ILT
CLIENT_SECRET:GEZSONRDNNHU3QSLGFDOLYSV3YSIB5KKM1C2IDNUKIHEYJIL


In [37]:
radius = 500 # Within a radius of 500 meters ...
limit = 100 # ... the top 100 venues

venues = []

for lat, long, post, borough, neighborhood in zip(toronto_central_df['Latitude'], 
                                                  toronto_central_df['Longitude'], 
                                                  toronto_central_df['PostalCode'], 
                                                  toronto_central_df['Borough'], 
                                                  toronto_central_df['Neighbourhood']):
    url = "https://api.foursquare.com/v2/venues/explore?client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}".format(
        CLIENT_ID,
        CLIENT_SECRET,
        VERSION,
        lat,
        long,
        radius, 
        limit)
    
    results = requests.get(url).json()["response"]['groups'][0]['items']
    
    for venue in results:
        venues.append((
            post, 
            borough,
            neighbourhood,
            lat, 
            long, 
            venue['venue']['name'], 
            venue['venue']['location']['lat'], 
            venue['venue']['location']['lng'],  
            venue['venue']['categories'][0]['name']))

In [40]:
venues_df = pd.DataFrame(venues)
venues_df.columns = ['PostalCode', 'Borough', 'Neighbourhood', 'BoroughLatitude', 'BoroughLongitude', 'VenueName', 'VenueLatitude', 'VenueLongitude', 'VenueCategory']
print(venues_df.head())
venues_df.shape

  PostalCode           Borough  \
0        M5A  Downtown Toronto   
1        M5A  Downtown Toronto   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M5A  Downtown Toronto   

                                       Neighbourhood  BoroughLatitude  \
0  Business reply mail Processing Centre, South C...          43.6555   
1  Business reply mail Processing Centre, South C...          43.6555   
2  Business reply mail Processing Centre, South C...          43.6555   
3  Business reply mail Processing Centre, South C...          43.6555   
4  Business reply mail Processing Centre, South C...          43.6555   

   BoroughLongitude               VenueName  VenueLatitude  VenueLongitude  \
0          -79.3626           Tandem Coffee      43.653559      -79.361809   
1          -79.3626        Roselle Desserts      43.653447      -79.362017   
2          -79.3626  Figs Breakfast & Lunch      43.655675      -79.364503   
3          -79.3626         The Yoga Lounge 

(1545, 9)

Analyze the venues dataframe. Size, venues per neighbourhood, and unique categories.

In [42]:
print(venues_df.shape)

(1545, 9)


In [43]:
print(venues_df.head())

  PostalCode           Borough  \
0        M5A  Downtown Toronto   
1        M5A  Downtown Toronto   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M5A  Downtown Toronto   

                                       Neighbourhood  BoroughLatitude  \
0  Business reply mail Processing Centre, South C...          43.6555   
1  Business reply mail Processing Centre, South C...          43.6555   
2  Business reply mail Processing Centre, South C...          43.6555   
3  Business reply mail Processing Centre, South C...          43.6555   
4  Business reply mail Processing Centre, South C...          43.6555   

   BoroughLongitude               VenueName  VenueLatitude  VenueLongitude  \
0          -79.3626           Tandem Coffee      43.653559      -79.361809   
1          -79.3626        Roselle Desserts      43.653447      -79.362017   
2          -79.3626  Figs Breakfast & Lunch      43.655675      -79.364503   
3          -79.3626         The Yoga Lounge 

In [47]:
print(venues_df.groupby(["PostalCode", "Borough", "Neighbourhood"]).count())

                                                                                BoroughLatitude  \
PostalCode Borough          Neighbourhood                                                         
M4E        East Toronto     Business reply mail Processing Centre, South Ce...                8   
M4K        East Toronto     Business reply mail Processing Centre, South Ce...               34   
M4L        East Toronto     Business reply mail Processing Centre, South Ce...               19   
M4M        East Toronto     Business reply mail Processing Centre, South Ce...                8   
M4N        Central Toronto  Business reply mail Processing Centre, South Ce...                2   
M4P        Central Toronto  Business reply mail Processing Centre, South Ce...                6   
M4R        Central Toronto  Business reply mail Processing Centre, South Ce...                4   
M4S        Central Toronto  Business reply mail Processing Centre, South Ce...               22   
M4T       

In [48]:
print('There are {} uniques categories.'.format(len(venues_df['VenueCategory'].unique())))

There are 215 uniques categories.


#### 5 Analyze Central Toronto

In [51]:
# one hot encoding
toronto_onehot = pd.get_dummies(venues_df[['VenueCategory']], prefix="", prefix_sep="")

# add postal, borough and neighborhood column back to dataframe
toronto_onehot['PostalCode'] = venues_df['PostalCode'] 
toronto_onehot['Borough'] = venues_df['Borough'] 
toronto_onehot['Neighbourhoods'] = venues_df['Neighbourhood'] 

# move postal, borough and neighborhood column to the first column
fixed_columns = list(toronto_onehot.columns[-3:]) + list(toronto_onehot.columns[:-3])
toronto_onehot = toronto_onehot[fixed_columns]

print(toronto_onehot.shape)
toronto_onehot.head()

(1545, 218)


PostalCode           Borough  \
0        M5A  Downtown Toronto   
1        M5A  Downtown Toronto   
2        M5A  Downtown Toronto   
3        M5A  Downtown Toronto   
4        M5A  Downtown Toronto   

                                      Neighbourhoods  Accessories Store  \
0  Business reply mail Processing Centre, South C...                  0   
1  Business reply mail Processing Centre, South C...                  0   
2  Business reply mail Processing Centre, South C...                  0   
3  Business reply mail Processing Centre, South C...                  0   
4  Business reply mail Processing Centre, South C...                  0   

   Afghan Restaurant  American Restaurant  Art Gallery  Art Museum  \
0                  0                    0            0           0   
1                  0                    0            0           0   
2                  0                    0            0           0   
3                  0                    0            0           0   
4                  0                    0            0           0   

   Arts & Crafts Store  Asian Restaurant  ...  Trail  Train Station  \
0                    0                 0  ...      0              0   
1                    0                 0  ...      0              0   
2                    0                 0  ...      0              0   
3                    0                 0  ...      0              0   
4                    0                 0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Whisky Bar  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0           0         0            0              0            0  
1           0         0            0              0            0  
2           0         0            0              0            0  
3           0         0            0              0            1  
4           0         0            0              0            0  

[5 rows x 218 columns]

Group rows by neighborhood and by taking the mean of the frequency of occurrence of each category.

In [53]:
toronto_grouped = toronto_onehot.groupby(["PostalCode", "Borough", "Neighbourhoods"]).mean().reset_index()

print(toronto_grouped.shape)
toronto_grouped.head()

(38, 218)


PostalCode          Borough  \
0        M4E     East Toronto   
1        M4K     East Toronto   
2        M4L     East Toronto   
3        M4M     East Toronto   
4        M4N  Central Toronto   

                                      Neighbourhoods  Accessories Store  \
0  Business reply mail Processing Centre, South C...                0.0   
1  Business reply mail Processing Centre, South C...                0.0   
2  Business reply mail Processing Centre, South C...                0.0   
3  Business reply mail Processing Centre, South C...                0.0   
4  Business reply mail Processing Centre, South C...                0.0   

   Afghan Restaurant  American Restaurant  Art Gallery  Art Museum  \
0                0.0             0.000000          0.0         0.0   
1                0.0             0.029412          0.0         0.0   
2                0.0             0.000000          0.0         0.0   
3                0.0             0.000000          0.0         0.0   
4                0.0             0.000000          0.0         0.0   

   Arts & Crafts Store  Asian Restaurant  ...     Trail  Train Station  \
0                  0.0               0.0  ...  0.125000            0.0   
1                  0.0               0.0  ...  0.029412            0.0   
2                  0.0               0.0  ...  0.000000            0.0   
3                  0.0               0.0  ...  0.000000            0.0   
4                  0.0               0.0  ...  0.000000            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                            0.0               0.0                    0.0   
1                            0.0               0.0                    0.0   
2                            0.0               0.0                    0.0   
3                            0.0               0.0                    0.0   
4                            0.0               0.0                    0.0   

   Whisky Bar  Wine Bar  Wings Joint  Women's Store  Yoga Studio  
0         0.0       0.0          0.0            0.0     0.000000  
1         0.0       0.0          0.0            0.0     0.029412  
2         0.0       0.0          0.0            0.0     0.000000  
3         0.0       0.0          0.0            0.0     0.000000  
4         0.0       0.0          0.0            0.0     0.000000  

[5 rows x 218 columns]

Print each neighborhood along with the top 20 most common venues.

In [55]:
num_top_venues = 20

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
areaColumns = ['PostalCode', 'Borough', 'Neighbourhoods']
freqColumns = []
for ind in np.arange(num_top_venues):
    try:
        freqColumns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        freqColumns.append('{}th Most Common Venue'.format(ind+1))
columns = areaColumns+freqColumns

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['PostalCode'] = toronto_grouped['PostalCode']
neighbourhoods_venues_sorted['Borough'] = toronto_grouped['Borough']
neighbourhoods_venues_sorted['Neighbourhoods'] = toronto_grouped['Neighbourhoods']

for ind in np.arange(toronto_grouped.shape[0]):
    row_categories = toronto_grouped.iloc[ind, :].iloc[3:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    neighbourhoods_venues_sorted.iloc[ind, 3:] = row_categories_sorted.index.values[0:num_top_venues]

print(neighbourhoods_venues_sorted.shape)
neighbourhoods_venues_sorted.head()

(38, 23)


PostalCode          Borough  \
0        M4E     East Toronto   
1        M4K     East Toronto   
2        M4L     East Toronto   
3        M4M     East Toronto   
4        M4N  Central Toronto   

                                      Neighbourhoods 1st Most Common Venue  \
0  Business reply mail Processing Centre, South C...                   Pub   
1  Business reply mail Processing Centre, South C...      Greek Restaurant   
2  Business reply mail Processing Centre, South C...             Pet Store   
3  Business reply mail Processing Centre, South C...        Baseball Field   
4  Business reply mail Processing Centre, South C...    Photography Studio   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0     Health Food Store                 Trail             Gastropub   
1        Ice Cream Shop    Italian Restaurant                  Café   
2      Sushi Restaurant            Steakhouse               Brewery   
3                  Park           Coffee Shop                 Diner   
4                  Park   Distribution Center  Fast Food Restaurant   

  5th Most Common Venue 6th Most Common Venue  7th Most Common Venue  ...  \
0          Neighborhood                Bakery            Cheese Shop  ...   
1            Restaurant           Yoga Studio         Cosmetics Shop  ...   
2         Burrito Place        Sandwich Place             Restaurant  ...   
3         Garden Center                   Gym  Performing Arts Venue  ...   
4        Farmers Market    Falafel Restaurant            Event Space  ...   

        11th Most Common Venue 12th Most Common Venue  \
0  Eastern European Restaurant      Electronics Store   
1                          Pub                 Lounge   
2                Movie Theater           Liquor Store   
3                      Dog Run      Electronics Store   
4          Dumpling Restaurant             Donut Shop   

        13th Most Common Venue 14th Most Common Venue 15th Most Common Venue  \
0         Ethiopian Restaurant            Event Space     Falafel Restaurant   
1                 Dessert Shop                    Spa           Liquor Store   
2           Light Rail Station   Fast Food Restaurant      Fish & Chips Shop   
3  Eastern European Restaurant    Dumpling Restaurant             Donut Shop   
4             Doner Restaurant                Dog Run         Discount Store   

  16th Most Common Venue   17th Most Common Venue 18th Most Common Venue  \
0         Farmers Market         Doner Restaurant         Discount Store   
1              Juice Bar  Fruit & Vegetable Store      Indian Restaurant   
2     Italian Restaurant           Ice Cream Shop             Board Shop   
3       Doner Restaurant     Ethiopian Restaurant            Yoga Studio   
4    Hawaiian Restaurant                    Diner           Dessert Shop   

  19th Most Common Venue 20th Most Common Venue  
0    Distribution Center      Fish & Chips Shop  
1                Brewery          Grocery Store  
2                    Gym             Beer Store  
3    Distribution Center         Discount Store  
4       Department Store          Deli / Bodega  

[5 rows x 23 columns]

#### 6 Cluster Neighborhoods

In [204]:
# set number of clusters
kclusters = 4

toronto_grouped_clustering = toronto_grouped.drop(["PostalCode", "Borough", "Neighbourhoods"], 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 2, 0, 1, 0, 1, 0])

In [207]:
neighbourhoods_venues_sorted.insert(0, 'Cluster_Labels', kmeans.labels_)

# create a new dataframe that includes the cluster as well as the top 20 venues for each neighborhood.
toronto_merged = toronto_central_df.copy()

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighbourhoods_venues_sorted.drop(["Borough", "Neighbourhoods"], 1).set_index("PostalCode"), on="PostalCode")

print(toronto_merged.shape)
toronto_merged

(39, 27)


PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M7A  Downtown Toronto   
2         M5B  Downtown Toronto   
3         M5C  Downtown Toronto   
4         M4E      East Toronto   
5         M5E  Downtown Toronto   
6         M5G  Downtown Toronto   
7         M6G  Downtown Toronto   
8         M5H  Downtown Toronto   
9         M6H      West Toronto   
10        M5J  Downtown Toronto   
11        M6J      West Toronto   
12        M4K      East Toronto   
13        M5K  Downtown Toronto   
14        M6K      West Toronto   
15        M4L      East Toronto   
16        M5L  Downtown Toronto   
17        M4M      East Toronto   
18        M4N   Central Toronto   
19        M5N   Central Toronto   
20        M4P   Central Toronto   
21        M5P   Central Toronto   
22        M6P      West Toronto   
23        M4R   Central Toronto   
24        M5R   Central Toronto   
25        M6R      West Toronto   
26        M4S   Central Toronto   
27        M5S  Downtown Toronto   
28        M6S      West Toronto   
29        M4T   Central Toronto   
30        M5T  Downtown Toronto   
31        M4V   Central Toronto   
32        M5V  Downtown Toronto   
33        M4W  Downtown Toronto   
34        M5W  Downtown Toronto   
35        M4X  Downtown Toronto   
36        M5X  Downtown Toronto   
37        M4Y  Downtown Toronto   
38        M7Y      East Toronto   

                                        Neighbourhood  Latitude  Longitude  \
0                           Regent Park, Harbourfront   43.6555   -79.3626   
1         Queen's Park, Ontario Provincial Government   43.6641   -79.3889   
2                            Garden District, Ryerson   43.6572   -79.3783   
3                                      St. James Town   43.6513   -79.3756   
4                                         The Beaches   43.6784   -79.2941   
5                                         Berczy Park   43.6456   -79.3754   
6                                  Central Bay Street   43.6564   -79.3860   
7                                            Christie   43.6683   -79.4205   
8                            Richmond, Adelaide, King   43.6496   -79.3833   
9                        Dufferin, Dovercourt Village   43.6655   -79.4378   
10  Harbourfront East, Union Station, Toronto Islands   43.6230   -79.3936   
11                           Little Portugal, Trinity   43.6480   -79.4177   
12                       The Danforth West, Riverdale   43.6803   -79.3538   
13           Toronto Dominion Centre, Design Exchange   43.6469   -79.3823   
14       Brockton, Parkdale Village, Exhibition Place   43.6383   -79.4301   
15                     India Bazaar, The Beaches West   43.6693   -79.3155   
16                     Commerce Court, Victoria Hotel   43.6492   -79.3823   
17                                    Studio District   43.6561   -79.3406   
18                                      Lawrence Park   43.7301   -79.3935   
19                                           Roselawn   43.7113   -79.4195   
20                                   Davisville North   43.7135   -79.3887   
21    Forest Hill North & West, Forest Hill Road Park   43.6966   -79.4120   
22                      High Park, The Junction South   43.6605   -79.4633   
23                  North Toronto West, Lawrence Park   43.7143   -79.4065   
24                The Annex, North Midtown, Yorkville   43.6736   -79.4035   
25                             Parkdale, Roncesvalles   43.6469   -79.4521   
26                                         Davisville   43.7020   -79.3853   
27                     University of Toronto, Harbord   43.6629   -79.3987   
28                                 Runnymede, Swansea   43.6512   -79.4828   
29                        Moore Park, Summerhill East   43.6899   -79.3853   
30          Kensington Market, Chinatown, Grange Park   43.6541   -79.3978   
31  Summerhill West, Rathnelly, South Hill, Forest...   43.6861   -79.4025   
32  CN Tower, King and Spadi

In [209]:
toronto_merged['marker color'] = pd.cut(toronto_merged['Cluster_Labels'], bins=4, 
                              labels=['red', 'blue', 'green', 'purple'])
toronto_merged                                      

PostalCode           Borough  \
0         M5A  Downtown Toronto   
1         M7A  Downtown Toronto   
2         M5B  Downtown Toronto   
3         M5C  Downtown Toronto   
4         M4E      East Toronto   
5         M5E  Downtown Toronto   
6         M5G  Downtown Toronto   
7         M6G  Downtown Toronto   
8         M5H  Downtown Toronto   
9         M6H      West Toronto   
10        M5J  Downtown Toronto   
11        M6J      West Toronto   
12        M4K      East Toronto   
13        M5K  Downtown Toronto   
14        M6K      West Toronto   
15        M4L      East Toronto   
16        M5L  Downtown Toronto   
17        M4M      East Toronto   
18        M4N   Central Toronto   
19        M5N   Central Toronto   
20        M4P   Central Toronto   
21        M5P   Central Toronto   
22        M6P      West Toronto   
23        M4R   Central Toronto   
24        M5R   Central Toronto   
25        M6R      West Toronto   
26        M4S   Central Toronto   
27        M5S  Downtown Toronto   
28        M6S      West Toronto   
29        M4T   Central Toronto   
30        M5T  Downtown Toronto   
31        M4V   Central Toronto   
32        M5V  Downtown Toronto   
33        M4W  Downtown Toronto   
34        M5W  Downtown Toronto   
35        M4X  Downtown Toronto   
36        M5X  Downtown Toronto   
37        M4Y  Downtown Toronto   
38        M7Y      East Toronto   

                                        Neighbourhood  Latitude  Longitude  \
0                           Regent Park, Harbourfront   43.6555   -79.3626   
1         Queen's Park, Ontario Provincial Government   43.6641   -79.3889   
2                            Garden District, Ryerson   43.6572   -79.3783   
3                                      St. James Town   43.6513   -79.3756   
4                                         The Beaches   43.6784   -79.2941   
5                                         Berczy Park   43.6456   -79.3754   
6                                  Central Bay Street   43.6564   -79.3860   
7                                            Christie   43.6683   -79.4205   
8                            Richmond, Adelaide, King   43.6496   -79.3833   
9                        Dufferin, Dovercourt Village   43.6655   -79.4378   
10  Harbourfront East, Union Station, Toronto Islands   43.6230   -79.3936   
11                           Little Portugal, Trinity   43.6480   -79.4177   
12                       The Danforth West, Riverdale   43.6803   -79.3538   
13           Toronto Dominion Centre, Design Exchange   43.6469   -79.3823   
14       Brockton, Parkdale Village, Exhibition Place   43.6383   -79.4301   
15                     India Bazaar, The Beaches West   43.6693   -79.3155   
16                     Commerce Court, Victoria Hotel   43.6492   -79.3823   
17                                    Studio District   43.6561   -79.3406   
18                                      Lawrence Park   43.7301   -79.3935   
19                                           Roselawn   43.7113   -79.4195   
20                                   Davisville North   43.7135   -79.3887   
21    Forest Hill North & West, Forest Hill Road Park   43.6966   -79.4120   
22                      High Park, The Junction South   43.6605   -79.4633   
23                  North Toronto West, Lawrence Park   43.7143   -79.4065   
24                The Annex, North Midtown, Yorkville   43.6736   -79.4035   
25                             Parkdale, Roncesvalles   43.6469   -79.4521   
26                                         Davisville   43.7020   -79.3853   
27                     University of Toronto, Harbord   43.6629   -79.3987   
28                                 Runnymede, Swansea   43.6512   -79.4828   
29                        Moore Park, Summerhill East   43.6899   -79.3853   
30          Kensington Market, Chinatown, Grange Park   43.6541   -79.3978   
31  Summerhill West, Rathnelly, South Hill, Forest...   43.6861   -79.4025   
32  CN Tower, King and Spadi

In [212]:
toronto_cluster1 = toronto_merged[toronto_merged['Cluster_Labels'] == 0]
toronto_cluster2 = toronto_merged[toronto_merged['Cluster_Labels'] == 1]
toronto_cluster3 = toronto_merged[toronto_merged['Cluster_Labels'] == 2]
toronto_cluster4 = toronto_merged[toronto_merged['Cluster_Labels'] == 3]

In [215]:
# create map
m = folium.Map(location=[43.653963, -79.387207], zoom_start=11)


# add markers cluster 1 to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster, color in zip(toronto_cluster1['Latitude'], toronto_cluster1['Longitude'], toronto_cluster1['PostalCode'], toronto_cluster1['Borough'], toronto_cluster1['Neighbourhood'], toronto_cluster1['Cluster Labels'], toronto_cluster1['marker color']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color='green',
        fill=True,
        fill_color='green',
        fill_opacity=0.7).add_to(m)

# add markers cluster 2 to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster, color in zip(toronto_cluster2['Latitude'], toronto_cluster2['Longitude'], toronto_cluster2['PostalCode'], toronto_cluster2['Borough'], toronto_cluster2['Neighbourhood'], toronto_cluster2['Cluster Labels'], toronto_cluster2['marker color']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color='pruple',
        fill=True,
        fill_color='purple',
        fill_opacity=0.7).add_to(m)
    
# add markers cluster 3 to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster, color in zip(toronto_cluster3['Latitude'], toronto_cluster3['Longitude'], toronto_cluster3['PostalCode'], toronto_cluster3['Borough'], toronto_cluster3['Neighbourhood'], toronto_cluster3['Cluster Labels'], toronto_cluster3['marker color']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color='red',
        fill=True,
        fill_color='red',
        fill_opacity=0.7).add_to(m)

# add markers cluster 4 to the map
markers_colors = []
for lat, lon, post, bor, poi, cluster, color in zip(toronto_cluster4['Latitude'], toronto_cluster4['Longitude'], toronto_cluster4['PostalCode'], toronto_cluster4['Borough'], toronto_cluster4['Neighbourhood'], toronto_cluster4['Cluster Labels'], toronto_cluster4['marker color']):
    label = folium.Popup('{} ({}): {} - Cluster {}'.format(bor, post, poi, cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='blue',
        fill_opacity=0.7).add_to(m)
    
       
m

Toronto city center is relatively homogenous with respect to the venues. 